# Project 1: Driving Licenses, Traffic Accidents and Casualties Analysis

## Problem Statment

Craft your problem statement here using the prompt from the README and other sources you find during your investigation.

## Executive Summary
Write an executive summary that summarizes the problem and your key findings.

### Contents:
- [Datasets Description](#Datasets-Description)
- [Data Import & Cleaning](#Data-Import-and-Cleaning)
- [Exploratory Data Analysis](#Exploratory-Data-Analysis)
- [Data Visualization](#Visualize-the-data)
- [Descriptive and Inferential Statistics](#Descriptive-and-Inferential-Statistics)
- [Outside Research](#Outside-Research)
- [Conclusions and Recommendations](#Conclusions-and-Recommendations)

**If you combine your problem statement, executive summary, data dictionary, and conclusions/recommendations, you have an amazing README.md file that quickly aligns your audience to the contents of your project.** Don't forget to cite your data sources!

*All libraries used should be added here, including any Jupyter magic commands*

In [1]:
#Imports
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

%matplotlib inline

## Datasets Description

[Driving Licenses](https://datasource.kapsarc.org/explore/dataset/saudi-arabia-driving-licenses-issued-in-the-kingdom-2004-2008/information/?disjunctive.administritive_area&sort=time_period&location=5,24.37495,45.08024&basemap=jawg.streets)
This dataset contains Saudi Arabia Driving Licenses Issued By Administrative Area for 1993 - 2016. Data from General Authority for Statistics . Follow datasource.kapsarc.org for timely data to advance energy economics research.

[Traffic Accidents and Casualties](https://datasource.kapsarc.org/explore/dataset/saudi-arabia-traffic-accidents-and-casualties-injured-dead-2008/export/?disjunctive.region&disjunctive.indicator&sort=time_period)
This dataset contains Saudi Arabia Traffic Accidents and Casualties by Region for 2016. Data from General Authority for Statistics. Follow datasource.kapsarc.org for timely data to advance energy economics research.


## Data Import and Cleaning

#### 1. Read In Driving Licenses & Traffic Accidents Data
Read in the `saudi-arabia-traffic-accidents-2008.csv` and `saudi-arabia-driving-licenses-2004-2008.csv` files and assign them to appropriately named pandas dataframes.

In [2]:
# Read in datasets from csv to pandas dataframes, load only useful columns
acc_columns = ['Year', 'Region', 'Indicator', 'Value', 'geo_point_2d']
lic_columns = ['Year', 'Administritive Area', 'Driving Liceses', 'geo_point_2d']

accidents = pd.read_csv('../data/saudi-arabia-traffic-accidents-2008.csv', sep=';', usecols=acc_columns)
licenses = pd.read_csv('../data/saudi-arabia-driving-licenses-2004-2008.csv', sep=';', usecols=lic_columns)

#### 2. Display data

Print the first 10 rows of each dataframe to your jupyter notebook

In [3]:
# Check dataframes using .head() & .tail() and .shape
accidents.head(10)

Year            Region                    Indicator   Value  \
0  (2016)            Makkah  No. of Casualties - Injured  12383n   
1  (2016)    Eastern Region     No. of Casualties - Dead   1113n   
2  (2016)            Tabouk     No. of Casualties - Dead    434n   
3  (2016)              Hail     No. of Casualties - Dead    299n   
4  (2016)              Hail             No. of Accidents  10199n   
5  (2016)  Northern Boarder     No. of Casualties - Dead    193n   
6  (2017)             Jazan             No. of Accidents  22161n   
7  (2017)           Al-Jouf             No. of Accidents  10469n   
8  (2017)             Total     No. of Casualties - Dead   7489n   
9  (2016)  Northern Boarder  No. of Casualties - Injured    631n   

                   geo_point_2d  
0   21.628962766, 41.2636898334  
1  23.2883307307, 50.1478302439  
2    27.90165499, 37.2598581394  
3   27.400845724, 41.4406960508  
4   27.400845724, 41.4406960508  
5   30.197074783, 41.8171186952  
6  17.2072313962, 42.6525807422  
7  29.9677445762, 39.5570371172  
8                           NaN  
9   30.197074783, 41.8171186952

In [4]:
accidents.tail()

Year    Region                    Indicator  Value  \
79  (2017)      Asir  No. of Casualties - Injured  2368n   
80  (2017)      Hail  No. of Casualties - Injured   773n   
81  (2017)  Al-Baaha     No. of Casualties - Dead   134n   
82  (2017)  Al-Baaha             No. of Accidents  1414n   
83  (2017)  Al-Baaha  No. of Casualties - Injured  1007n   

                    geo_point_2d  
79   19.1387847401, 42.941470036  
80   27.400845724, 41.4406960508  
81  20.1484058503, 41.4665484892  
82  20.1484058503, 41.4665484892  
83  20.1484058503, 41.4665484892

In [5]:
accidents.shape

(84, 5)

In [6]:
licenses.head(10)

Year Administritive Area Driving Liceses                 geo_point_2d
0  (2012)              Makkah         143954n  21.628962766, 41.2636898334
1  (1994)              Makkah          75349n  21.628962766, 41.2636898334
2  (1998)              Makkah          86082n  21.628962766, 41.2636898334
3  (2004)              Makkah          86636n  21.628962766, 41.2636898334
4  (2005)              Makkah          65099n  21.628962766, 41.2636898334
5  (2008)              Makkah          77013n  21.628962766, 41.2636898334
6  (2014)             Madinah          31447n  24.927534539, 39.4918237173
7  (2013)             Madinah          28513n  24.927534539, 39.4918237173
8  (2012)             Madinah          14459n  24.927534539, 39.4918237173
9  (1994)             Madinah           3767n  24.927534539, 39.4918237173

In [7]:
licenses.tail()

Year Administritive Area Driving Liceses                 geo_point_2d
345  (2000)              Tabouk           8473n   27.90165499, 37.2598581394
346  (2006)              Tabouk           8259n   27.90165499, 37.2598581394
347  (2013)                Hail           9731n  27.400845724, 41.4406960508
348  (1997)                Hail           6606n  27.400845724, 41.4406960508
349  (1999)                Hail           3139n  27.400845724, 41.4406960508

In [8]:
licenses.shape

(350, 4)

#### 3. Briefly describe the data

Take your time looking through the data and briefly describe the data in the markdown cell below. Note things about what the columns might mean, and the general information that is conveyed in the dataframe.

Answer:
- The accidents dataframe consists of yearly data (2016-2017) about vehicle accidents for regions in KSA, with various variables and values presented in a long format, also has a geo point column
- The licenses dataframe consists of yearly data over 25 years about the number of licenses issued per region in KSA, also has a geo point column as in accidents

#### 4a. How complete is the data?

Investigate missing values etc.

In [9]:
accidents.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 84 entries, 0 to 83
Data columns (total 5 columns):
Year            84 non-null object
Region          84 non-null object
Indicator       84 non-null object
Value           84 non-null object
geo_point_2d    78 non-null object
dtypes: object(5)
memory usage: 3.4+ KB


In [10]:
licenses.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 350 entries, 0 to 349
Data columns (total 4 columns):
Year                   350 non-null object
Administritive Area    350 non-null object
Driving Liceses        350 non-null object
geo_point_2d           325 non-null object
dtypes: object(4)
memory usage: 11.0+ KB


### The geopoint column seems to be missing some entries in both accidents & licenses. All other observations seem to be nun-null with 84 & 350 entries respectively for other columns in accidents and licenses

#### 4b. Are there any obvious issues with the observations?

**What is the minimum *possible* value for each dataset? What is the maximum *possible* value?**

In [11]:
print(accidents.min())
print(accidents.max())

Year                   (2016)
Region               Al-Baaha
Indicator    No. of Accidents
Value                   1007n
dtype: object
Year                              (2017)
Region                             Total
Indicator    No. of Casualties - Injured
Value                             99925n
dtype: object


In [12]:
print(licenses.min())
print(licenses.max())

Year                    (1993)
Administritive Area    Al-Baha
Driving Liceses        100291n
dtype: object
Year                   (2017)
Administritive Area     Total
Driving Liceses         9983n
dtype: object


Data cannot be properly interpreted because of data type mismatch

In [13]:
accidents.describe()

Year Region         Indicator    Value                 geo_point_2d
count       84     84                84       84                           78
unique       2     14                 3       84                           13
top     (2017)  Total  No. of Accidents  145541n  24.927534539, 39.4918237173
freq        42      6                28        1                            6

In [14]:
accidents.Indicator.unique()

array(['No. of Casualties - Injured', 'No. of Casualties - Dead',
       'No. of Accidents'], dtype=object)

In [15]:
# check year range for dataset
accidents.Year.value_counts()

(2017)    42
(2016)    42
Name: Year, dtype: int64

In [102]:
licenses.describe()

year  driving_licenses
count   325.000000        325.000000
mean   2005.000000      28915.347692
std       7.222222      47419.058819
min    1993.000000        915.000000
25%    1999.000000       4846.000000
50%    2005.000000      10267.000000
75%    2011.000000      25423.000000
max    2017.000000     495307.000000

In [17]:
# check year range for dataset
licenses.Year.value_counts().sort_index()

(1993)    14
(1994)    14
(1995)    14
(1996)    14
(1997)    14
(1998)    14
(1999)    14
(2000)    14
(2001)    14
(2002)    14
(2003)    14
(2004)    14
(2005)    14
(2006)    14
(2007)    14
(2008)    14
(2009)    14
(2010)    14
(2011)    14
(2012)    14
(2013)    14
(2014)    14
(2015)    14
(2016)    14
(2017)    14
Name: Year, dtype: int64

#### 5. What are your data types? 
Display the data types of each feature. 

In [18]:
# check data types with .dtypes()
accidents.dtypes

Year            object
Region          object
Indicator       object
Value           object
geo_point_2d    object
dtype: object

In [19]:
licenses.dtypes

Year                   object
Administritive Area    object
Driving Liceses        object
geo_point_2d           object
dtype: object

What did you learn?
- Do any of them seem odd?  
- Which ones are not as they should be?  

Answer:

- all columns for both datasets are of dtype object which usually means they are in string form
- Year could be numeric or a datetime, Region/Area a category, and driving 'liceses' & most values as numeric data after correcting them

#### 6. Fix incorrect data types
Based on what you discovered above, use appropriate methods to re-type incorrectly typed data.
- Define a function that will allow you to convert numerical columns to an appropriate numeric type. Use `map` or `apply` to change these columns in each dataframe.

In [20]:
def col_to_numeric(x):
    try:
        num_x = int(x)
        return num_x
    except:
        print('UNSUCCESFUL, CANNOT BE CONVERTED TO FLOAT')

- Make new columns `x` and `y` using `geo_point_2d`

In [21]:
# get x and y geo columns
accidents['x'] = accidents['geo_point_2d'].str.replace(' ', '').apply(lambda x: str(x)[:str(x).find(',')])
accidents['y'] = accidents['geo_point_2d'].str.replace(' ', '').apply(lambda y: str(y)[str(y).find(',')+1:])
accidents.head()

Year          Region                    Indicator   Value  \
0  (2016)          Makkah  No. of Casualties - Injured  12383n   
1  (2016)  Eastern Region     No. of Casualties - Dead   1113n   
2  (2016)          Tabouk     No. of Casualties - Dead    434n   
3  (2016)            Hail     No. of Casualties - Dead    299n   
4  (2016)            Hail             No. of Accidents  10199n   

                   geo_point_2d              x              y  
0   21.628962766, 41.2636898334   21.628962766  41.2636898334  
1  23.2883307307, 50.1478302439  23.2883307307  50.1478302439  
2    27.90165499, 37.2598581394    27.90165499  37.2598581394  
3   27.400845724, 41.4406960508   27.400845724  41.4406960508  
4   27.400845724, 41.4406960508   27.400845724  41.4406960508

In [22]:
licenses['x'] = licenses['geo_point_2d'].str.replace(' ', '').apply(lambda x: str(x)[:str(x).find(',')])
licenses['y'] = licenses['geo_point_2d'].str.replace(' ', '').apply(lambda y: str(y)[str(y).find(',')+1:])
licenses.head()

Year Administritive Area Driving Liceses                 geo_point_2d  \
0  (2012)              Makkah         143954n  21.628962766, 41.2636898334   
1  (1994)              Makkah          75349n  21.628962766, 41.2636898334   
2  (1998)              Makkah          86082n  21.628962766, 41.2636898334   
3  (2004)              Makkah          86636n  21.628962766, 41.2636898334   
4  (2005)              Makkah          65099n  21.628962766, 41.2636898334   

              x              y  
0  21.628962766  41.2636898334  
1  21.628962766  41.2636898334  
2  21.628962766  41.2636898334  
3  21.628962766  41.2636898334  
4  21.628962766  41.2636898334

- Fix any individual values preventing other columns from being the appropriate type.

In [23]:
accidents.Region.unique(), licenses['Administritive Area'].unique()

(array(['Makkah', 'Eastern Region', 'Tabouk', 'Hail', 'Northern Boarder',
        'Jazan', 'Al-Jouf', 'Total', 'Al-Baaha', 'Riyadh', 'Madinah',
        'Qassim', 'Najran', 'Asir'], dtype=object),
 array(['Makkah', 'Madinah', 'Al-Qaseem', 'Eastern', 'Assir', 'Tabouk',
        'Hail', 'Northern Boarder', 'Jazan', 'Najran', 'Al-Baha',
        'Al-Jouf', 'Total', 'Riyadh'], dtype=object))

In [24]:
# fix value issues including typos and differing names between dfs
accidents['Year'] = accidents['Year'].str.replace('(', '').str.replace(')', '')
accidents['Value'] = accidents['Value'].str.replace('n', '')
accidents['Region'] = accidents['Region'].str.replace('Boarder', 'Border')
accidents['Region'] = accidents['Region'].str.replace('Assir', 'Asir')
accidents.head()

Year          Region                    Indicator  Value  \
0  2016          Makkah  No. of Casualties - Injured  12383   
1  2016  Eastern Region     No. of Casualties - Dead   1113   
2  2016          Tabouk     No. of Casualties - Dead    434   
3  2016            Hail     No. of Casualties - Dead    299   
4  2016            Hail             No. of Accidents  10199   

                   geo_point_2d              x              y  
0   21.628962766, 41.2636898334   21.628962766  41.2636898334  
1  23.2883307307, 50.1478302439  23.2883307307  50.1478302439  
2    27.90165499, 37.2598581394    27.90165499  37.2598581394  
3   27.400845724, 41.4406960508   27.400845724  41.4406960508  
4   27.400845724, 41.4406960508   27.400845724  41.4406960508

In [25]:
licenses['Year'] = licenses['Year'].str.replace('(', '').str.replace(')', '')
licenses['Driving Liceses'] = licenses['Driving Liceses'].str.replace('n', '')
licenses['Administritive Area'] = licenses['Administritive Area'].str.replace('Boarder', 'Border')
licenses['Administritive Area'] = licenses['Administritive Area'].str.replace('Assir', 'Asir')
licenses['Administritive Area'] = licenses['Administritive Area'].str.replace('Al-Baha', 'Al-Baaha')
licenses['Administritive Area'] = licenses['Administritive Area'].str.replace('Eastern', 'Eastern Region')
licenses['Administritive Area'] = licenses['Administritive Area'].str.replace('Al-Qaseem', 'Qassim')
licenses.head()

Year Administritive Area Driving Liceses                 geo_point_2d  \
0  2012              Makkah          143954  21.628962766, 41.2636898334   
1  1994              Makkah           75349  21.628962766, 41.2636898334   
2  1998              Makkah           86082  21.628962766, 41.2636898334   
3  2004              Makkah           86636  21.628962766, 41.2636898334   
4  2005              Makkah           65099  21.628962766, 41.2636898334   

              x              y  
0  21.628962766  41.2636898334  
1  21.628962766  41.2636898334  
2  21.628962766  41.2636898334  
3  21.628962766  41.2636898334  
4  21.628962766  41.2636898334

- Finish your data modifications by making sure the columns are now typed appropriately.

In [26]:
# convert to numeric
accidents['Year'] = accidents.Year.apply(col_to_numeric)
accidents['Value'] = accidents.Value.apply(col_to_numeric)
accidents['Region'] = accidents.Region.astype('category')
licenses['Year'] = licenses.Year.apply(col_to_numeric)
licenses['Driving Liceses'] = licenses['Driving Liceses'].apply(col_to_numeric)
licenses['Administritive Area'] = licenses['Administritive Area'].astype('category')
accidents.head()

Year          Region                    Indicator  Value  \
0  2016          Makkah  No. of Casualties - Injured  12383   
1  2016  Eastern Region     No. of Casualties - Dead   1113   
2  2016          Tabouk     No. of Casualties - Dead    434   
3  2016            Hail     No. of Casualties - Dead    299   
4  2016            Hail             No. of Accidents  10199   

                   geo_point_2d              x              y  
0   21.628962766, 41.2636898334   21.628962766  41.2636898334  
1  23.2883307307, 50.1478302439  23.2883307307  50.1478302439  
2    27.90165499, 37.2598581394    27.90165499  37.2598581394  
3   27.400845724, 41.4406960508   27.400845724  41.4406960508  
4   27.400845724, 41.4406960508   27.400845724  41.4406960508

In [27]:
licenses.head()

Year Administritive Area  Driving Liceses                 geo_point_2d  \
0  2012              Makkah           143954  21.628962766, 41.2636898334   
1  1994              Makkah            75349  21.628962766, 41.2636898334   
2  1998              Makkah            86082  21.628962766, 41.2636898334   
3  2004              Makkah            86636  21.628962766, 41.2636898334   
4  2005              Makkah            65099  21.628962766, 41.2636898334   

              x              y  
0  21.628962766  41.2636898334  
1  21.628962766  41.2636898334  
2  21.628962766  41.2636898334  
3  21.628962766  41.2636898334  
4  21.628962766  41.2636898334

- Display the data types again to confirm they are correct.

In [28]:
accidents.dtypes

Year               int64
Region          category
Indicator         object
Value              int64
geo_point_2d      object
x                 object
y                 object
dtype: object

In [29]:
licenses.dtypes

Year                      int64
Administritive Area    category
Driving Liceses           int64
geo_point_2d             object
x                        object
y                        object
dtype: object

#### 7. Rename columns
Change the names of the columns to more expressive names so that you can tell the difference the Driving Licenses  columns and the & Traffic Accidents columns. Your solution should map all column names being changed at once (no repeated singular name-changes). **We will be combining these two datasets, and so you should name columns in an appropriate way**.

**Guidelines**:
- Column names should be all lowercase (you will thank yourself when you start pushing data to SQL later in the course)
- Column names should not contain spaces (underscores will suffice--this allows for using the `df.column_name` method to access columns in addition to `df['column_name']`.
- Column names should be unique and informative (the only feature that we actually share between dataframes is the state).
- Please also fix any typos you see in the column names.

In [30]:
# check columns
list(accidents.columns.values), list(licenses.columns.values)

(['Year', 'Region', 'Indicator', 'Value', 'geo_point_2d', 'x', 'y'],
 ['Year', 'Administritive Area', 'Driving Liceses', 'geo_point_2d', 'x', 'y'])

In [31]:
# fix column names and change admin area to region in licenses df to unify naming across dfs
accidents.columns = ['year', 'region', 'indicator', 'value', 'geo_point_2d', 'geo_point_lat', 'geo_point_long']
licenses.columns =['year', 'region', 'driving_licenses', 'geo_point_2d',  'geo_point_lat', 'geo_point_long']

In [32]:
accidents.head()

year          region                    indicator  value  \
0  2016          Makkah  No. of Casualties - Injured  12383   
1  2016  Eastern Region     No. of Casualties - Dead   1113   
2  2016          Tabouk     No. of Casualties - Dead    434   
3  2016            Hail     No. of Casualties - Dead    299   
4  2016            Hail             No. of Accidents  10199   

                   geo_point_2d  geo_point_lat geo_point_long  
0   21.628962766, 41.2636898334   21.628962766  41.2636898334  
1  23.2883307307, 50.1478302439  23.2883307307  50.1478302439  
2    27.90165499, 37.2598581394    27.90165499  37.2598581394  
3   27.400845724, 41.4406960508   27.400845724  41.4406960508  
4   27.400845724, 41.4406960508   27.400845724  41.4406960508

In [33]:
licenses.head()

year  region  driving_licenses                 geo_point_2d geo_point_lat  \
0  2012  Makkah            143954  21.628962766, 41.2636898334  21.628962766   
1  1994  Makkah             75349  21.628962766, 41.2636898334  21.628962766   
2  1998  Makkah             86082  21.628962766, 41.2636898334  21.628962766   
3  2004  Makkah             86636  21.628962766, 41.2636898334  21.628962766   
4  2005  Makkah             65099  21.628962766, 41.2636898334  21.628962766   

  geo_point_long  
0  41.2636898334  
1  41.2636898334  
2  41.2636898334  
3  41.2636898334  
4  41.2636898334

In [34]:
# check license totals to infer whether licenses column is total issued licenses or total new licenses issued on that year
licenses[licenses.region == 'Total'].sort_values('year')

year region  driving_licenses geo_point_2d geo_point_lat geo_point_long
60   1993  Total            294893          NaN            na            nan
61   1994  Total            303520          NaN            na            nan
240  1995  Total            251878          NaN            na            nan
316  1996  Total            237623          NaN            na            nan
315  1997  Total            289296          NaN            na            nan
317  1998  Total            338350          NaN            na            nan
145  1999  Total            332210          NaN            na            nan
163  2000  Total            273238          NaN            na            nan
147  2001  Total            264725          NaN            na            nan
146  2002  Total            282038          NaN            na            nan
62   2003  Total            254858          NaN            na            nan
148  2004  Total            268738          NaN            na            nan
149  2005  Total            251203          NaN            na            nan
63   2006  Total            267508          NaN            na            nan
164  2007  Total            331356          NaN            na            nan
165  2008  Total            330158          NaN            na            nan
150  2009  Total            376275          NaN            na            nan
143  2010  Total            386510          NaN            na            nan
162  2011  Total            400558          NaN            na            nan
144  2012  Total            411204          NaN            na            nan
239  2013  Total            565314          NaN            na            nan
142  2014  Total            625414          NaN            na            nan
58   2015  Total            435439          NaN            na            nan
314  2016  Total            697017          NaN            na            nan
59   2017  Total            928165          NaN            na            nan

##### Numbers seem to indicate that it refers to new licenses issued on that year

#### 8. Create a data dictionary

Now that we've fixed our data, and given it appropriate names, let's create a [data dictionary](http://library.ucmerced.edu/node/10249). 

A data dictionary provides a quick overview of features/variables/columns, alongside data types and descriptions. The more descriptive you can be, the more useful this document is.

Example of a Fictional Data Dictionary Entry: 

|Feature|Type|Dataset|Description|
|---|---|---|---|
|**county_pop**|*integer*|2010 census|The population of the county (units in thousands, where 2.5 represents 2500 people).| 
|**per_poverty**|*float*|2010 census|The percent of the county over the age of 18 living below the 200% of official US poverty rate (units percent to two decimal places 98.10 means 98.1%)|

[Here's a quick link to a short guide for formatting markdown in Jupyter notebooks](https://jupyter-notebook.readthedocs.io/en/stable/examples/Notebook/Working%20With%20Markdown%20Cells.html).

Provided is the skeleton for formatting a markdown table, with columns headers that will help you create a data dictionary to quickly summarize your data, as well as some examples. **This would be a great thing to copy and paste into your custom README for this project.**

In [35]:
accidents.dtypes, licenses.dtypes

(year                 int64
 region            category
 indicator           object
 value                int64
 geo_point_2d        object
 geo_point_lat       object
 geo_point_long      object
 dtype: object, year                   int64
 region              category
 driving_licenses       int64
 geo_point_2d          object
 geo_point_lat         object
 geo_point_long        object
 dtype: object)

|Feature|Type|Dataset|Description|
|---|---|---|---|
|year|int|Driving_Licenses/Traffic_Accidents|The year in which the observation happened, only 2 years in the accidents df, 25 years in the licenses| 
|region|category|Driving_Licenses/Traffic_Accidents|The administrative region in which the observation was conducted, 13 regions in KSA + country 'Total'| 
|indicator|object|Traffic_Accidents|The variable being measured in the observation, measures relating to accident and casualty numbers| 
|value|int|Traffic_Accidents|The integer value associated with the adjacent indicator|
|driving_licenses|int|Driving_Licenses|The number of driving licenses issued in each region category by year|
|geo_point_2d|object|Driving_Licenses/Traffic_Accidents|Representation of latitude and longitude of the geolocation of the region| 
|geo_point_lat|object|Traffic_Accidents|Represents latitude of regions geolocation| 
|geo_point_long|object|Traffic_Accidents|Represents longitude of regions geolocation| 

#### 9. Drop unnecessary rows

This study concerns accident and license in regions/area. Please do the followings:
0. Which rows are not associated with regions?
1. Confirm the numbers of all regions add up to total.
2. Remove the rows that are not associated with regions

In [36]:
# Any row with 'Total' as a region is unnecessary, let's check if totals are correct
# Step1: get totals for each year via groupby sum
acc_no_total = accidents[accidents.region != 'Total']
acc_no_total.groupby(['year', 'indicator']).sum()

value
year indicator                          
2016 No. of Accidents             533380
     No. of Casualties - Dead       9031
     No. of Casualties - Injured   38120
2017 No. of Accidents             460488
     No. of Casualties - Dead       7489
     No. of Casualties - Injured   33199

In [37]:
# Step2: compare with 'Total' rows
accidents[accidents.region == "Total"][['year', 'region', 'indicator', 'value']].sort_values(['year', 'indicator'])

year region                    indicator   value
29  2016  Total             No. of Accidents  533380
77  2016  Total     No. of Casualties - Dead    9031
30  2016  Total  No. of Casualties - Injured   38120
68  2017  Total             No. of Accidents  460488
8   2017  Total     No. of Casualties - Dead    7489
69  2017  Total  No. of Casualties - Injured   33199

In [38]:
# do the same for licenses df
lic_no_total = licenses[licenses.region != 'Total']
lic_sum_totals = lic_no_total.groupby('year').sum()
lic_sum_totals.head()

driving_licenses
year                  
1993            294893
1994            303520
1995            251878
1996            237623
1997            289296

In [39]:
lic_totals = licenses[licenses.region == 'Total'].set_index('year').sort_index()[['driving_licenses']]
lic_totals.head()

driving_licenses
year                  
1993            294893
1994            303520
1995            251878
1996            237623
1997            289296

In [40]:
# test both dfs are equal
lic_sum_totals.equals(lic_totals)

True

### We asserted that the 'total' rows are equivalent to summing up the regions observations per year and therefore is redundant information that we can safely remove

In [41]:
# remove 'total' rows and confirm 6 rows removed from accidents and 25 from licenses
accidents = accidents[accidents.region != 'Total']
licenses = licenses[licenses.region != 'Total']
accidents.shape, licenses.shape

((78, 7), (325, 6))

#### 10a. Data formats
Is the License data in long or wide format? How about the Accident data? And why?

In [42]:
accidents.head(1)

year  region                    indicator  value  \
0  2016  Makkah  No. of Casualties - Injured  12383   

                  geo_point_2d geo_point_lat geo_point_long  
0  21.628962766, 41.2636898334  21.628962766  41.2636898334

In [43]:
licenses.head(1)

year  region  driving_licenses                 geo_point_2d geo_point_lat  \
0  2012  Makkah            143954  21.628962766, 41.2636898334  21.628962766   

  geo_point_long  
0  41.2636898334

### Accident data is in long format thanks to indicator and value columns while license data is in wide format

#### 10b. Changing formats
Which format would work best for both dataframes? Transform the datasets to the same format before merging in the next task.

In [44]:
# Wide format would work best so we will transform accidents to wide
acc_pivoted = accidents.pivot_table(index=['year', 'region', 'geo_point_2d', 'geo_point_lat', 'geo_point_long'], columns="indicator", values='value')

In [45]:
# format the new df and drop geo_point_2d column
acc_pivoted.columns.name = None
acc_pivot_columns_order = ['year', 'region', 'No. of Accidents', 'No. of Casualties - Dead', 'No. of Casualties - Injured', 'geo_point_lat', 'geo_point_long']
acc_pivoted = acc_pivoted.reset_index()[acc_pivot_columns_order]
acc_pivoted.head(2)

year    region  No. of Accidents  No. of Casualties - Dead  \
0  2016  Al-Baaha              3618                       185   
1  2016   Al-Jouf             10212                       245   

   No. of Casualties - Injured  geo_point_lat geo_point_long  
0                         1227  20.1484058503  41.4665484892  
1                          920  29.9677445762  39.5570371172

In [46]:
# rename columns to lowercase conventions
acc_pivot_updated_columns = ['year', 'region', 'no_accidents', 'casualties_dead', 'casualties_injured', 'geo_point_lat', 'geo_point_long']
acc_pivoted.columns = acc_pivot_updated_columns
acc_pivoted

year           region  no_accidents  casualties_dead  casualties_injured  \
0   2016         Al-Baaha          3618              185                1227   
1   2016          Al-Jouf         10212              245                 920   
2   2016             Asir         32684              983                2541   
3   2016   Eastern Region        101619             1113                4893   
4   2016             Hail         10199              299                 932   
5   2016            Jazan         25873              514                2259   
6   2016          Madinah         14475              822                4184   
7   2016           Makkah        134838             2243               12383   
8   2016           Najran          3235              173                 706   
9   2016  Northern Border         13196              193                 631   
10  2016           Qassim         25664              397                1513   
11  2016           Riyadh        141736             1430                4207   
12  2016           Tabouk         16031              434                1724   
13  2017         Al-Baaha          1414              134                1007   
14  2017          Al-Jouf         10469              240                 862   
15  2017             Asir         25245              828                2368   
16  2017   Eastern Region         81433              963                4073   
17  2017             Hail          9379              265                 773   
18  2017            Jazan         22161              466                2207   
19  2017          Madinah         16070              580                2917   
20  2017           Makkah        145541             1641               10623   
21  2017           Najran          3193              194                 766   
22  2017  Northern Border         11540              112                 374   
23  2017           Qassim         22405              319                1446   
24  2017           Riyadh         99925             1399                4540   
25  2017           Tabouk         11713              348                1243   

    geo_point_lat geo_point_long  
0   20.1484058503  41.4665484892  
1   29.9677445762  39.5570371172  
2   19.1387847401   42.941470036  
3   23.2883307307  50.1478302439  
4    27.400845724  41.4406960508  
5   17.2072313962  42.6525807422  
6    24.927534539  39.4918237173  
7    21.628962766  41.2636898334  
8   18.4511003068  46.8688495726  
9    30.197074783  41.8171186952  
10  26.1967063704  43.3429998455  
11  23.0814341634  45.5857512766  
12    27.90165499  37.2598581394  
13  20.1484058503  41.4665484892  
14  29.9677445762  39.5570371172  
15  19.1387847401   42.941470036  
16  23.2883307307  50.1478302439  
17   27.400845724  41.4406960508  
18  17.2072313962  42.6525807422  
19   24.927534539  39.4918237173  
20   21.628962766  41.2636898334  
21  18.4511003068  46.8688495726  
22   30.197074783  41.8171186952  
23  26.1967063704  43.3429998455  
24  23.0814341634  45.5857512766  
25    27.90165499  37.2598581394

#### 11. Merge dataframes

Join the Driving Licenses & Traffic Accidents dataframes using the **region** and **year** in each dataframe as the key. Assign this to a new variable. Which join should you use to preserve the most data?

In [47]:
# to preserve the most data we do an outer or left merge in this case
licenses.drop('geo_point_2d', axis=1).merge(acc_pivoted, on=['region', 'year', 'geo_point_lat', 'geo_point_long'], how='outer').sort_values('year').iloc[:, [0, 1, 2, 5, 6, 7, 3, 4]]

year           region  driving_licenses  no_accidents  casualties_dead  \
207  1993           Najran              3769           NaN              NaN   
192  1993           Tabouk              3174           NaN              NaN   
313  1993             Asir              6226           NaN              NaN   
167  1993           Makkah             65977           NaN              NaN   
34   1993  Northern Border              1510           NaN              NaN   
197  1993             Hail             10484           NaN              NaN   
126  1993         Al-Baaha              3270           NaN              NaN   
300  1993           Qassim             73618           NaN              NaN   
38   1993            Jazan              6009           NaN              NaN   
307  1993   Eastern Region              9964           NaN              NaN   
228  1993          Madinah              4122           NaN              NaN   
137  1993           Riyadh            103886           NaN              NaN   
216  1993          Al-Jouf              2884           NaN              NaN   
9    1994          Madinah              3767           NaN              NaN   
314  1994             Asir              5185           NaN              NaN   
260  1994  Northern Border              1457           NaN              NaN   
97   1994           Tabouk              3282           NaN              NaN   
1    1994           Makkah             75349           NaN              NaN   
52   1994         Al-Baaha              2729           NaN              NaN   
83   1994           Qassim             68916           NaN              NaN   
281  1994           Najran              3880           NaN              NaN   
116  1994            Jazan              3826           NaN              NaN   
253  1994             Hail             11131           NaN              NaN   
139  1994           Riyadh            110545           NaN              NaN   
238  1994   Eastern Region             10918           NaN              NaN   
134  1994          Al-Jouf              2535           NaN              NaN   
45   1995           Najran              3268           NaN              NaN   
74   1995          Madinah              4155           NaN              NaN   
202  1995            Jazan              2284           NaN              NaN   
217  1995          Al-Jouf              2270           NaN              NaN   
..    ...              ...               ...           ...              ...   
250  2015             Hail              7172           NaN              NaN   
155  2015           Najran              1864           NaN              NaN   
151  2015            Jazan             13931           NaN              NaN   
78   2015           Qassim             13631           NaN              NaN   
317  2016           Tabouk             13610       16031.0            434.0   
235  2016   Eastern Region            142307      101619.0           1113.0   
243  2016             Asir             41367       32684.0            983.0   
50   2016         Al-Baaha              4614        3618.0            185.0   
265  2016            Jazan             11504       25873.0            514.0   
143  2016           Makkah            136055      134838.0           2243.0   
196  2016             Hail             54507       10199.0            299.0   
170  2016          Madinah             24540       14475.0            822.0   
289  2016          Al-Jouf              9371       10212.0            245.0   
174  2016           Qassim              9180       25664.0            397.0   
107  2016  Northern Border              4958       13196.0            193.0   
272  2016           Riyadh            242851      141736.0           1430.0   
119  2016           Najran              2153        3235.0            173.0   
80   2017           Qassim              6318       22405.0            319.0   
285  2017         Al-Baaha              4131        141

In [48]:
# in this case since we would like to compare licenses and accidents data we'll do an inner or right merge and organize
combined = licenses.drop('geo_point_2d', axis=1).merge(acc_pivoted, on=['region', 'year', 'geo_point_lat', 'geo_point_long'], how='inner').iloc[:, [0, 1, 2, 5, 6, 7, 3, 4]]
combined = combined.sort_values(['year', 'region']).reset_index(drop=True)
combined

year           region  driving_licenses  no_accidents  casualties_dead  \
0   2016         Al-Baaha              4614          3618              185   
1   2016          Al-Jouf              9371         10212              245   
2   2016             Asir             41367         32684              983   
3   2016   Eastern Region            142307        101619             1113   
4   2016             Hail             54507         10199              299   
5   2016            Jazan             11504         25873              514   
6   2016          Madinah             24540         14475              822   
7   2016           Makkah            136055        134838             2243   
8   2016           Najran              2153          3235              173   
9   2016  Northern Border              4958         13196              193   
10  2016           Qassim              9180         25664              397   
11  2016           Riyadh            242851        141736             1430   
12  2016           Tabouk             13610         16031              434   
13  2017         Al-Baaha              4131          1414              134   
14  2017          Al-Jouf              8091         10469              240   
15  2017             Asir             39133         25245              828   
16  2017   Eastern Region            126816         81433              963   
17  2017             Hail             22035          9379              265   
18  2017            Jazan             18345         22161              466   
19  2017          Madinah             36165         16070              580   
20  2017           Makkah            142487        145541             1641   
21  2017           Najran             13825          3193              194   
22  2017  Northern Border              5280         11540              112   
23  2017           Qassim              6318         22405              319   
24  2017           Riyadh            495307         99925             1399   
25  2017           Tabouk             10232         11713              348   

    casualties_injured  geo_point_lat geo_point_long  
0                 1227  20.1484058503  41.4665484892  
1                  920  29.9677445762  39.5570371172  
2                 2541  19.1387847401   42.941470036  
3                 4893  23.2883307307  50.1478302439  
4                  932   27.400845724  41.4406960508  
5                 2259  17.2072313962  42.6525807422  
6                 4184   24.927534539  39.4918237173  
7                12383   21.628962766  41.2636898334  
8                  706  18.4511003068  46.8688495726  
9                  631   30.197074783  41.8171186952  
10                1513  26.1967063704  43.3429998455  
11                4207  23.0814341634  45.5857512766  
12                1724    27.90165499  37.2598581394  
13                1007  20.1484058503  41.4665484892  
14                 862  29.9677445762  39.5570371172  
15                2368  19.1387847401   42.941470036  
16                4073  23.2883307307  50.1478302439  
17                 773   27.400845724  41.4406960508  
18                2207  17.2072313962  42.6525807422  
19                2917   24.927534539  39.4918237173  
20               10623   21.628962766  41.2636898334  
21                 766  18.4511003068  46.8688495726  
22                 374   30.197074783  41.8171186952  
23                1446  26.1967063704  43.3429998455  
24                4540  23.0814341634  45.5857512766  
25                1243    27.90165499  37.2598581394

#### 12. Save your cleaned, merged dataframe

Use a relative path to save out your data as `combined_datasets.csv`.

In [49]:
combined.to_csv('../data/combined_datasets.csv', index=False)

## Exploratory Data Analysis


### Summary Statistics
Transpose the output of pandas `describe` method to create a quick overview of each numeric feature.

In [87]:
combined.describe().transpose()

count          mean            std     min       25%  \
year                 26.0   2016.500000       0.509902  2016.0   2016.00   
driving_licenses     26.0  62507.000000  107293.917762  2153.0   8363.25   
no_accidents         26.0  38225.692308   46493.705035  1414.0  10276.25   
casualties_dead      26.0    635.384615     549.783381   112.0    241.25   
casualties_injured   26.0   2743.038462    2922.641127   374.0    923.00   

                        50%       75%       max  
year                 2016.5   2017.00    2017.0  
driving_licenses    16085.0  51222.00  495307.0  
no_accidents        16050.5  30981.25  145541.0  
casualties_dead       415.5    929.25    2243.0  
casualties_injured   1618.5   3784.00   12383.0

#### Manually calculate standard deviation

$$\sigma = \sqrt{\frac{1}{n}\sum_{i=1}^n(x_i - \mu)^2}$$

- Write a function to calculate standard deviation using the formula above

In [79]:
# two methods, changed n to n-1 in second to get sample std not population
def std_dev_pop(series):
    return np.sqrt(((series - series.mean())**2).sum()/len(series))

def std_dev_sample(series):
    return np.sqrt(((series - series.mean())**2).sum()/(len(series)-1))

In [85]:
# same as in .describe()
std_dev_sample(combined.driving_licenses)

107293.9177620055

In [82]:
std_dev_pop(combined.driving_licenses)

105210.34237659337

In [83]:
np.std(combined.driving_licenses)

105210.34237659337

- Use a **dictionary comprehension** to apply your standard deviation function to each numeric column in the dataframe.  **No loops**  
- Assign the output to variable `sd` as a dictionary where: 
    - Each column name is now a key 
    - That standard deviation of the column is the value 
     
*Example Output :* `{'Driving_Licenses_x': 120, 'Traffic_Accidents_x': 120, ...}`

In [94]:
sd = {column:std_dev_sample(combined[column]) for column in combined[['driving_licenses', 'no_accidents', 'casualties_dead', 'casualties_injured']].columns}
sd

{'driving_licenses': 107293.9177620055,
 'no_accidents': 46493.70503521459,
 'casualties_dead': 549.7833811182784,
 'casualties_injured': 2922.6411272103765}

Do your manually calculated standard deviations match up with the output from pandas `describe`? What about numpy's `std` method?

Answer:

matches up with describe() because using sample std, would match with np.std() if population std

#### Investigate trends in the data
Using sorting and/or masking (along with the `.head` method to not print our entire dataframe), consider the following questions:

- Which regions have the highest and lowest Driving Licenses based on years (1993 - 2017)
- Which regions have the highest and lowest mean Traffic Accidents numbers for years (2016 - 2017)
- What are the regions that have more Driving Licenses issued in each year than the year average?
- What are the regions that have more Traffic Accidents happened in each year than the year average?

Based on what you've just observed, have you identified any regions that you're especially interested in? **Make a note of these and state *why* you think they're interesting**.

**You should comment on your findings at each step in a markdown cell below your code block**. Make sure you include at least one example of sorting your dataframe by a column, and one example of using boolean filtering (i.e., masking) to select a subset of the dataframe.

In [96]:
licenses.head(2)

year  region  driving_licenses                 geo_point_2d geo_point_lat  \
0  2012  Makkah            143954  21.628962766, 41.2636898334  21.628962766   
1  1994  Makkah             75349  21.628962766, 41.2636898334  21.628962766   

  geo_point_long  
0  41.2636898334  
1  41.2636898334

In [103]:
# lowest and highest Driving Licenses based on years (1993 - 2017)
licenses[(licenses['driving_licenses'] == licenses.driving_licenses.max()) | (licenses['driving_licenses'] == licenses.driving_licenses.min())]

year  region  driving_licenses                  geo_point_2d  \
210  2015  Tabouk               915    27.90165499, 37.2598581394   
292  2017  Riyadh            495307  23.0814341634, 45.5857512766   

     geo_point_lat geo_point_long  
210    27.90165499  37.2598581394  
292  23.0814341634  45.5857512766

In [117]:
# highest and lowest mean Traffic Accidents numbers for years (2016 - 2017)
combined[(combined['no_accidents'].isin(combined.groupby('year')['no_accidents'].max())) | (combined['no_accidents'].isin(combined.groupby('year')['no_accidents'].min()))]

year    region  driving_licenses  no_accidents  casualties_dead  \
8   2016    Najran              2153          3235              173   
11  2016    Riyadh            242851        141736             1430   
13  2017  Al-Baaha              4131          1414              134   
20  2017    Makkah            142487        145541             1641   

    casualties_injured  geo_point_lat geo_point_long  
8                  706  18.4511003068  46.8688495726  
11                4207  23.0814341634  45.5857512766  
13                1007  20.1484058503  41.4665484892  
20               10623   21.628962766  41.2636898334

In [139]:
# What are the regions that have more Driving Licenses issued in each year than the year average?
mean_lic = licenses.groupby('year').mean().driving_licenses.values
mean_lic = dict(enumerate(mean_lic, 1993))
mean_lic = licenses['year'].map(mean_lic)
licenses[licenses['driving_licenses'] > mean_lic].sort_values('year')

year          region  driving_licenses                  geo_point_2d  \
186  1993          Makkah             65977   21.628962766, 41.2636898334   
152  1993          Riyadh            103886  23.0814341634, 45.5857512766   
325  1993          Qassim             73618  26.1967063704, 43.3429998455   
1    1994          Makkah             75349   21.628962766, 41.2636898334   
154  1994          Riyadh            110545  23.0814341634, 45.5857512766   
89   1994          Qassim             68916  26.1967063704, 43.3429998455   
153  1995          Riyadh            101608  23.0814341634, 45.5857512766   
187  1995          Makkah             60356   21.628962766, 41.2636898334   
14   1995          Qassim             49690  26.1967063704, 43.3429998455   
156  1996          Riyadh            110131  23.0814341634, 45.5857512766   
15   1996          Qassim             42693  26.1967063704, 43.3429998455   
299  1996          Makkah             46195   21.628962766, 41.2636898334   
74   1997          Makkah             76623   21.628962766, 41.2636898334   
70   1997          Riyadh             92145  23.0814341634, 45.5857512766   
260  1997  Eastern Region             58814  23.2883307307, 50.1478302439   
294  1998          Riyadh            114962  23.0814341634, 45.5857512766   
334  1998  Eastern Region             70650  23.2883307307, 50.1478302439   
2    1998          Makkah             86082   21.628962766, 41.2636898334   
333  1999  Eastern Region             64973  23.2883307307, 50.1478302439   
66   1999          Riyadh            100291  23.0814341634, 45.5857512766   
75   1999          Makkah            103535   21.628962766, 41.2636898334   
76   2000          Makkah             87053   21.628962766, 41.2636898334   
243  2000          Riyadh             45345  23.0814341634, 45.5857512766   
198  2000  Eastern Region             65767  23.2883307307, 50.1478302439   
188  2001          Makkah             85593   21.628962766, 41.2636898334   
199  2001  Eastern Region             74119  23.2883307307, 50.1478302439   
318  2002          Makkah             88593   21.628962766, 41.2636898334   
295  2002          Riyadh             25423  23.0814341634, 45.5857512766   
197  2002  Eastern Region             85035  23.2883307307, 50.1478302439   
319  2003          Makkah             74236   21.628962766, 41.2636898334   
..    ...             ...               ...                           ...   
5    2008          Makkah             77013   21.628962766, 41.2636898334   
202  2008  Eastern Region            104626  23.2883307307, 50.1478302439   
245  2008          Riyadh             36798  23.0814341634, 45.5857512766   
262  2009  Eastern Region            108544  23.2883307307, 50.1478302439   
298  2009          Riyadh             38014  23.0814341634, 45.5857512766   
78   2009          Makkah            112045   21.628962766, 41.2636898334   
87   2010          Qassim             36737  26.1967063704, 43.3429998455   
241  2010          Riyadh             31378  23.0814341634, 45.5857512766   
73   2010          Makkah            113531   21.628962766, 41.2636898334   
257  2010  Eastern Region            110295  23.2883307307, 50.1478302439   
246  2011          Makkah            110647   21.628962766, 41.2636898334   
95   2011  Eastern Region            151087  23.2883307307, 50.1478302439   
0    2012          Makkah            143954   21.628962766, 41.2636898334   
96   2012  Eastern Region            129971  23.2883307307, 50.1478302439   
69   2013          Makkah            131108   21.628962766, 41.2636898334   
65   2013          Riyadh            165994  23.0814341634, 45.5857512766   
195  2013  Eastern Region            123325  23.2883307307, 50.1478302439   
155  2014          Riyadh            194208  23.0814341634, 45.5857512766   
72   2014          Makkah            149234   21.628962766, 41.2636898334   
255  2014  Eastern Region            121774  23.2883307307, 50.1478302439   
6

In [141]:
# What are the regions that have more Traffic Accidents happened in each year than the year average?
mean_acc = combined.groupby('year').mean().no_accidents.values
mean_acc = dict(enumerate(mean_acc, 2016))
mean_acc = combined['year'].map(mean_acc)
combined[combined['no_accidents'] > mean_acc].sort_values('year')

year          region  driving_licenses  no_accidents  casualties_dead  \
3   2016  Eastern Region            142307        101619             1113   
7   2016          Makkah            136055        134838             2243   
11  2016          Riyadh            242851        141736             1430   
16  2017  Eastern Region            126816         81433              963   
20  2017          Makkah            142487        145541             1641   
24  2017          Riyadh            495307         99925             1399   

    casualties_injured  geo_point_lat geo_point_long  
3                 4893  23.2883307307  50.1478302439  
7                12383   21.628962766  41.2636898334  
11                4207  23.0814341634  45.5857512766  
16                4073  23.2883307307  50.1478302439  
20               10623   21.628962766  41.2636898334  
24                4540  23.0814341634  45.5857512766

## Visualize the data

There's not a magic bullet recommendation for the right number of plots to understand a given dataset, but visualizing your data is *always* a good idea. Not only does it allow you to quickly convey your findings (even if you have a non-technical audience), it will often reveal trends in your data that escaped you when you were looking only at numbers.

Some recommendations on plotting:
- Plots have titles
- Plots have axis labels
- Plots have appropriate tick labels
- All text is legible in a plot
- Plots demonstrate meaningful and valid relationships
- Plots are interpreted to aid understanding

There is such a thing as too many plots, and there are a *lot* of bad plots. You might make some! (But hopefully not with the guided prompts below).

#### Use Seaborn's heatmap with pandas `.corr()` to visualize correlations between all numeric features

Heatmaps are generally not appropriate for presentations, and should often be excluded from reports as they can be visually overwhelming. **However**, they can be extremely useful in identify relationships of potential interest (as well as identifying potential collinearity before modeling).

*example*:
```python
sns.heatmap(df.corr())
```

Please take time to format your output, adding a title. Look through some of the additional arguments and options. (Axis labels aren't really necessary, as long as the title is informative).

#### Define a custom function to subplot histograms

We should write a function that will take the names of 2+ columns and subplot histograms. While you can use pandas plotting or Seaborn here, matplotlib gives you greater control over all aspects of your plots.

[Helpful Link for Plotting Multiple Figures](https://matplotlib.org/users/pyplot_tutorial.html#working-with-multiple-figures-and-axes)

Here's some starter code:

In [54]:
def subplot_histograms(dataframe, list_of_columns, list_of_titles, list_of_xlabels):
    nrows = int(np.ceil(len(list_of_columns)/2) # Makes sure you have enough rows
    fig, ax = plt.subplots(nrows=nrows, ncols=2) # You'll want to specify your figsize
    ax = ax.ravel() # Ravel turns a matrix into a vector, which is easier to iterate
    for i, column in enumerate(list_of_columns): # Gives us an index value to get into all our lists
        ax[i].hist(dataframe[column]) # feel free to add more settings
        # Set titles, labels, etc here for each subplot

SyntaxError: invalid syntax (<ipython-input-54-f8881e15e6cc>, line 3)

#### Plot and interpret histograms 
For each of the following:
- Driving Licenses number
- Traffic Accidents number

Please plot two additional histograms that will help with your understanding of the data

In [ ]:
# Code

#### Plot and interpret scatter plots

For each of the following:
- Driving Licenses vs. Traffic Accidents for 2017
- Driving Licenses vs. Traffic Accidents for 2016
- Driving Licenses for 2016 vs 2017
- Traffic Accidents for 2016 vs 2017

Plot the two variables against each other using matplotlib or Seaborn

Your plots should show:
- Two clearly labeled axes
- A proper title
- Using colors and symbols that are clear and unmistakable

**Feel free to write a custom function, and subplot if you'd like.** Functions save both time and space.


In [ ]:
# Code

#### Plot and interpret boxplots

For each numeric variable in the dataframe create a boxplot using Seaborn. Boxplots demonstrate central tendency and spread in variables. In a certain sense, these are somewhat redundant with histograms, but you may be better able to identify clear outliers or differences in IQR, etc.

Multiple values can be plotted to a single boxplot as long as they are of the same relative scale (meaning they have similar min/max values).

Each boxplot should:
- Only include variables of a similar scale
- Have clear labels for each variable
- Have appropriate titles and labels

In [ ]:
# Code

#### Feel free to do additional plots below
*(do research and choose your own chart types & variables)*

Are there any additional trends or relationships you haven't explored? Was there something interesting you saw that you'd like to dive further into? It's likely that there are a few more plots you might want to generate to support your narrative and recommendations that you are building toward. **As always, make sure you're interpreting your plots as you go**.

## Descriptive and Inferential Statistics

#### Summarizing Distributions

Above, we used pandas `describe` to provide quick summary statistics of our numeric columns. We also demonstrated many visual relationships.

As data scientists, having a complete understanding of data is imperative prior to modeling.

While we will continue to build our analytic tools, we know that measures of *central tendency*, *spread*, and *shape/skewness* provide a quick summary of distributions.

For each variable in your data, summarize the underlying distributions (in words & statistics)
 - Be thorough in your verbal description of these distributions.
 - Be sure to back up these summaries with statistics.

Answers:

#### We generally assuming that data we sample from a population will be normally distributed. Do we observe this trend?

Answer:

Does This Assumption Hold for:
    - Driving Licenses
    - Traffic Accidents
Explain your answers for each distribution and how you think this will affect estimates made from these data.

Answer:

#### Statistical Evaluation of Distributions 

**If you feel it's appropriate**, using methods we discussed in class, run hypothesis tests to compare variables of interest in our dataset. 

In [ ]:
# Code

## Outside Research

Based upon your observations, choose **three** regions that demonstrate interesting trends in the number of driving licenses and traffic accidents. Spend some time doing outside research on provincial and central policies that might influence these rates, and summarize your findings below. **Feel free to go back and create new plots that highlight these states of interest**. If you bring in any outside tables or charts, make sure you are explicit about having borrowed them. If you quote any text, make sure that it renders as being quoted. (Make sure that you cite your sources -- check with you local instructor for citation preferences).

Answer:

## Conclusions and Recommendations

- Based on your exploration of the data, what are you key takeaways and recommendations?
- Are there additional data you desire that would better inform your investigations?

Answer: